src/orion/client/runner.py

```diff
@@ -0,0 +1,323 @@
+# pylint:disable=too-many-instance-attributes
+"""
+Runner
+======
+
+Executes the optimization process
+"""
+import logging
+import time
+from contextlib import contextmanager
+from dataclasses import dataclass
+
+import orion.core
+from orion.core.utils.exceptions import (
+    BrokenExperiment,
+    CompletedExperiment,
+    InvalidResult,
+    ReservationRaceCondition,
+    ReservationTimeout,
+    WaitingForTrials,
+)
+from orion.core.utils.flatten import flatten, unflatten
+from orion.core.worker.consumer import ExecutionError
+from orion.core.worker.trial import AlreadyReleased
+from orion.executor.base import AsyncException, AsyncResult
+
+
+class LazyWorkers(Exception):
+    """Raised when all the workers have been idle for a given amount of time"""
+
+    pass
+
+
+def _optimize(trial, fct, trial_arg, **kwargs):
+    """Execute a trial on a worker"""
+    kwargs.update(flatten(trial.params))
+
+    if trial_arg:
+        kwargs[trial_arg] = trial
+
+    return fct(**unflatten(kwargs))
+
+
+log = logging.getLogger(__name__)
+
+
+@contextmanager
+def _timer(self, name):
+    start = time.time()
+    yield
+    total = time.time() - start
+
+    value = getattr(self, name)
+    setattr(self, name, value + total)
+
+
+@dataclass
+class _Stat:
+    sample: int = 0
+    scatter: int = 0
+    gather: int = 0
+
+    def time(self, name):
+        """Measure elapsed  time of a given block"""
+        return _timer(self, name)
+
+    def report(self):
+        """Show the elapsed time of different blocks"""
+        lines = [
+            f"Sample  {self.sample:7.4f}",
+            f"Scatter {self.scatter:7.4f}",
+            f"Gather  {self.gather:7.4f}",
+        ]
+        return "\n".join(lines)
+
+
+class Runner:
+    """Run the optimization process given the current executor"""
+
+    def __init__(
+        self,
+        client,
+        fct,
+        pool_size,
+        idle_timeout,
+        max_trials_per_worker,
+        max_broken,
+        trial_arg,
+        on_error,
+        interrupt_signal_code=None,
+        **kwargs,
+    ):
+        self.client = client
+        self.fct = fct
+        self.pool_size = pool_size
+        self.max_trials_per_worker = max_trials_per_worker
+        self.max_broken = max_broken
+        self.trial_arg = trial_arg
+        self.on_error = on_error
+        self.kwargs = kwargs
+
+        self.gather_timeout = 0.01
+        self.idle_timeout = idle_timeout
+
+        self.worker_broken_trials = 0
+        self.trials = 0
+        self.futures = []
+        self.pending_trials = dict()
+        self.free_worker = pool_size
+        self.stat = _Stat()
+
+        if interrupt_signal_code is None:
+            interrupt_signal_code = orion.core.config.worker.interrupt_signal_code
+
+        self.interrupt_signal_code = interrupt_signal_code
+
+    @property
+    def is_done(self):
+        """Returns true if the experiment has finished."""
+        return self.client.is_done
```

Comment:
Maybe this should be more specific to the runner. It will also depend on `max_trials_per_worker`, so the runner could be done before the experiment is.

Reply:
We already have `has_remaining()` for that, makes `is_running` easier to understand
Currently each is_running condition is represented by a single property that represent a single condition.
All those conditions are combined in is_running alone.

So each condition are easy to understand and the combination is easy to understand if we mix conditions more it will start to become less straight forward

Reply:
Sounds good!